fairlearn/metrics/_plotter.py

```diff
@@ -0,0 +1,289 @@
+_CONF_INTERVALS_MUST_BE_ARRAY = "Calculated conf_intervals must be array of length 2"
+_ERRORS_NEGATIVE_VALUE_ERROR = "Calculated relative errors must be positive"
+_CONF_INTERVALS_FLIPPED_BOUNDS_ERROR = \
+    "Calculated conf_intervals' upper bound cannot be less than lower bound"
+
+
+def _is_arraylike(error):
+    return (isinstance(error, np.ndarray) or isinstance(error, list))
+
+
+def _check_if_metrics_and_error_metrics_same_length(metrics, errors, conf_intervals):
+    if errors is not None:
+        if len(errors) != len(metrics):
+            raise ValueError(_METRIC_AND_ERRORS_NOT_SAME_LENGTH_ERROR)
+    elif conf_intervals is not None:
+        if len(conf_intervals) != len(metrics):
+            raise ValueError(_METRIC_AND_CONF_INTERVALS_NOT_SAME_LENGTH_ERROR)
+
+
+def _check_if_metrics_length_zero(metrics):
+    if len(metrics) == 0:
+        raise ValueError(_METRIC_LENGTH_ZERO_ERROR)
+
+
+def _check_for_ambiguous_error_metric(errors, conf_intervals):
+    # ambiguous error metric if both are provided
+    if errors is not None and conf_intervals is not None:
+        raise ValueError(_GIVEN_BOTH_ERRORS_AND_CONF_INT_ERROR)
+
+
+def _check_for_valid_metrics_format(metrics):
+    # ensure metrics is either list, str, or None
+    if not (isinstance(metrics, list) or isinstance(metrics, str) or metrics is None):
+        raise ValueError(_METRICS_NOT_LIST_OR_STR_ERROR.format(type(metrics)))
+
+
+def _convert_scalar_error_to_array_error(df_errors):
+    if all((isinstance(error, float) or isinstance(error, int)) for error in df_errors):
+        return df_errors.apply(lambda x: [x, x])
+    if all((_is_arraylike(error) and len(error) == 2) for error in df_errors):
+        return df_errors
+    else:
+        raise ValueError(_ERRORS_MUST_BE_SCALAR_OR_ARRAY)
+
+
+def _check_valid_errors(df_errors):
+    for tup in df_errors:
+        if not all(ele >= 0 for ele in tup):
+            raise ValueError(_ERRORS_NEGATIVE_VALUE_ERROR)
+
+
+def _check_valid_conf_interval(df_conf_intervals):
+    for tup in df_conf_intervals:
+        if not _is_arraylike(tup) or len(tup) != 2:
+            raise ValueError(_CONF_INTERVALS_MUST_BE_ARRAY)
+        if tup[0] > tup[1]:
+            raise ValueError(_CONF_INTERVALS_FLIPPED_BOUNDS_ERROR)
+
+
+def _plot_df(df, metrics, kind, subplots, legend_label, df_all_errors=None, **kwargs):
+    if df_all_errors is not None:
+        yerr = np.array(
+            [np.array([[*row] for row in df_all_errors[metric]]).T for metric in metrics]
+        )
+        if kind == "scatter":
+            axs = df[metrics].plot(linestyle='', marker='o', yerr=yerr,
+                                   subplots=subplots, **kwargs)
+        else:
+            axs = df[metrics].plot(kind=kind, yerr=yerr, subplots=subplots, **kwargs)
+
+        if isinstance(axs, np.ndarray):
+            for ax in axs.flatten():
+                if kind == "scatter":
```

Comment:
The body of this `for` looks to be very similar to the `else` clause. Can a subroutine be extracted?